### PR TITLE
Fix HP draining in non-play sections

### LINKED ISSUE #7607: HP drains in non-play sections
**Describe the bug:** In osu!lazer, HP drain occurs in non-play sections, like right after a break or after skipping the beginning of the song but before the first note was hit.

This may be an issue that occurs in sections like before a break but after the last note, but I have not tested that yet.

**Screenshots or videos showing encountered issue:** https://youtu.be/lzn6P9MOdG4
Notice that osu!lazer immediately drains HP after the break, but osu!stable only drains upon hitting the first note.

**osu!lazer version:** master branch 52b66fdc167059ca0336dde474e2ee84073d04b1

### PR DESCRIPTION
**Pull request by kornkaobat**
Fixes #7607.

### CODE REVIEW

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (LittleEndu) [inline osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:50]: Why is this an integer? It's only ever incremented and checked if above zero. Could be bool instead.

Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:50]: It is so that I can keep track of Hits using `breakendcounter++;`

Comment (iiSaLMaN) [inline osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:50]:
```suggestion
        private bool judgedAfterBreak;
```
is probably what you want

Comment (LittleEndu) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:50]: That's what I said tho. You only write to this value by resetting it or incrementing it. And you only read from this value by seeing if it has been incremented. idk how is it not clear that you can make this a boolean instead.

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (iiSaLMaN) [inline osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:70]:
```csharp
if (IsBreakTime.Value || !judgedAfterBreak)
{
    judgedAfterBreak = false;
    return;
}

// draining logic
...
```

Comment (iiSaLMaN) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:70]: also provide a comment as to why draining will only be applied when a hit object is judged post-break.

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: Fixed in 61be2f5.

Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: Fixed ( naming is different ) in 61be2f5.

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: !judgedAfterBreak and judgedAfterBreak = false is the same thing or am I wrong here?

Comment (LittleEndu) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: ! is faster than ==

Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: I think I would only add `return;` because I can't seem to grasp the` ...|| !judgedAfterBreak and judgedAfterBreak = false` part.

Comment (LittleEndu) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: You can get to the assignment via the other variable in the or operator

Comment (TheWildTree) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: He means that there's a simple oversight in the suggestion, making it essentially

```csharp
if (!judgedAfterBreak)
    judgedAfterBreak = false;
```

Should've been without the `!` here.

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (iiSaLMaN) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: `breakend` doesn't make much sense when talking about draining health after first hit object being hit (we're at the beginning, not after a break end).

Also please follow our naming style for consistency and better code readability (each word from second onwards starts with capital letter): 
`Judged after break` field for example becomes `judgedAfterBreak` not `judgedafterbreak`

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (iiSaLMaN) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: @kornkaobat I've suggested it for better code readability:

current code: 
1. if on break time, clear `judgedAfterBreak`. 
2. if not on break time and is `judgedAfterBreak` then do draining logic.

suggested code: 
1. if on break time or not `judgedAfterBreak` then clear `judgedAfterBreak` and don't do draining logic (aka `return;`). 
2. do draining logic

also clearing `judgedAfterBreak` inside the check is intended because we're featuring `IsBreakTime` on the check and don't really like breaking it to 2 separate `if`s.

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (bdach) [inline osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:72]: The `!IsBreakTime.Value` check is now redundant, as the method early-returns if it's true (so it can't *not* be false at this point).

After eliminating it you could as well negate the other check, and I *think* the following variant is equivalent:
```csharp
// can check this first, as drain is re-enabled externally by ApplyResultInternal
if (!applyDrain)
    return;

// drain is on - check if a break is starting; if yes, disable and early-return
if (IsBreakTime.Value)
{
    applyDrain = false;
    return;
}

// actually apply drain
```

**Review by bdach — CHANGES_REQUESTED** [review]
This should have tests. `TestSceneDrainingHealthProcessor` already exists, so it shouldn't be very hard to add a test for this behaviour.

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: Fixed in e09786e.

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: Fixed in e09786e.

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: Fixed via bdach's suggestion e09786e.

Comment (mcendu) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: Why aren't you using the "resolve conversation" button?

Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs]: Already done that.

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (smoogipoo) [inline osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:100]: A similar change needs to be done to `computeDrainRate()` below (change probably needs to happen to line 144), which simulates this whole `Update()` + `ApplyResult` cycle.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by smoogipoo** [comment]
As I mentioned in https://github.com/ppy/osu/issues/8585#issuecomment-608208468, this also needs to account for the section _before_ the break, which also shouldn't drain.

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (smoogipoo) [inline osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:81]: No.

**Review by smoogipoo — CHANGES_REQUESTED** [review]
I believe this could be done by computing all the drain sections inside `ApplyBeatmap()` and checking if the current time lies within any of those sections. It's a similar implementation to how `IsBreakTime` is computed.

File: osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs
Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:81]: Can you explain the input-output of this method ApplyBeatmap(IBeatmap beatmap)?

Comment (smoogipoo) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:81]: It receives in a beatmap, and it extracts whatever it needs out of it. It does whatever it wants to do.

Comment (mcendu) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:81]: Maybe calculate something like `List<System.Tuple<int,int>> NonDrainSections`?

Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:81]: /// <summary>
        /// The minimum duration required for a break to have any effect.
        /// </summary>
        public const double MIN_BREAK_DURATION = 650;

Comment (kornkaobat) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:81]: I think we should approach this problem by changing broken beatmap Break Starttimes because, we can't predict into the future and assure if the Current Hitobject that we're hitting is the closest Hitobject to the upcoming Breaktime.

Comment (iiSaLMaN) [reply osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs:81]: @mcendu's idea is probably what you need to solve this (although the tuple must be of `double,double` because that's the type we use to store time values and also must be a `private` local). 
where you want to calculate this would be inside the `ApplyBeatmap(beatmap)` method, now think on how to calculate it, search off what `Beatmap` has. (should be fairly simple)

**Comment by bdach** [comment]
Please, please, *please* do not push code that you know to be crashing.

**Comment by kornkaobat** [comment]
Why is that though, does the CI waste bandwidth or something else? It won't be merged yet so what's the harm.

**Comment by LittleEndu** [comment]
Bandwidth isn't the only resource CI is wasting when it has to test commits that one knows to fail.

**Comment by kornkaobat** [comment]
Ok then.

**Comment by bdach** [comment]
It does waste CI bandwidth (other PR builds won't run), and in general development by trial-and-error demonstrates the lack of necessary knowledge to actually fix the problem, and also a lack of knowledge of the *process* of git workflow (PRs are supposed to be functional code always, unless opened as draft, waiting for a prerequisite merge, or just plain mistake - those happen).

**Comment by kornkaobat** [comment]
How to catch what specific error when a Unhandled Exception happens while osu! is running and is about to crash.

**Comment by bdach** [comment]
Check log files for the stack trace, or just use an IDE and run in debug mode (most do automatic breakpoints on exceptions)...?

With all due respect, this sort of question you should be looking to find out the answer to on your own. The discussion on this PR has been inordinately long for what should be a relatively simple change. We cannot possibly provide this level of counsel if you choose to continue submitting code.

**Comment by iiSaLMaN** [comment]
@kornkaobat Please understand what @smoogipoo's review tells you https://github.com/ppy/osu/pull/8542#pullrequestreview-387197443, otherwise you can't proceed with incorrect code like this, If you don't really know what he means then search up, understand, learn how to do it and attempt until you reach to the correct solution. 

Otherwise if you still can't figure it out then it would be better to just close this PR and let us solve it in the correct way.

**Comment by kornkaobat** [comment]
Ok I'll leave it to you guys and girls to solve it yourselves then.

**Comment by kornkaobat** [comment]
I've solved this problem by doing some hardcore programming. Enjoy.

**Comment by bdach** [comment]
The latest commit is so illegible and wrong that instead of spending even more time on explaining basic programming practices I am going to go ahead and close this again. Please do not reopen this unless you are sure what you are doing.

**Comment by kornkaobat** [comment]
It does work so what's wrong with it?

**Comment by bdach** [comment]
If you insist:

* Ten added fields. I *really* don't think that many is necessary.
* Checking hit object type by getting it as string and *comparing to a magic string*.
* Duplicating the same 12-line snippet in two places.
* And the change in general is just unmaintainable. I don't understand what the code is doing with any *hashes* (which, as far as I can see, are not hashes) and I don't think I should have to attempt to.

I would strongly recommend starting with the basics of C# and object-oriented programming design before attempting to submit code again.

**Comment by kornkaobat** [comment]
Ok then live with your inferior programs. I want to see if you guys and girls can do any better. Trust me you'll run into problems when you try to code this, and I'll not be there to help. This game release cycle will suffer because this is a core gameplay problem and you'll be remembering me but, I'm far ahead in life I wouldn't came back and look at this repo again. Farewell.

**Pull request closed without merging**
